### PR TITLE
solve_saveat

### PR DESCRIPTION
**Pull request by pnkraemer**

### CODE REVIEW

File: tests/test_solve_saveat.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        f=f, df=df, saveat=saveat, u0=u0, solver=solver, atol=atol, rtol=rtol
+    )
+    # Each of those is of the form ((rv, backward_model), t)
+    state_terminal, solution_stacked, state_initial = solution
+
+    # Uhmmm, each output (terminal, stacked, initial) looks so innocent,
+    # but they are _packed_ full of information.
+    # Long list of assertions incoming...
+    # This probably implies that the output needs some refactoring,
+    # but for now I am happy with it, because it does the job.
```
Comment (pnkraemer) [inline tests/test_solve_saveat.py:50]: it might be more usable if we stack this into a suitable namedtuple, for example. But so far, I like to keep it so naked, and I'd say we can iterate if it bothers anyone too much?

File: tornadox/solve.py [hunk 2]
```diff
@@ -59,6 +59,53 @@ def solve_ivp_for_terminal_value(*, f, df, tspan, u0, solver, **solver_kwargs):
     return extract_qoi_fn(t, state)
 
 
+@partial(jax.jit, static_argnames=("f", "df", "solver"))
+def solve_ivp_saveat(*, f, df, u0, solver, saveat, **solver_kwargs):
+    # assume t0 and tmax are the "boundary" elements in saveat.
+    # As such, `saveat` makes `tspan` obsolete
+    # Also assume it is sorted.
+    # If not, anything could happen.
```
Comment (pnkraemer) [inline tornadox/solve.py:67]: this could be a docstring, at least at some point (soon) :smile:

File: tornadox/solve.py [hunk 3]
```diff
@@ -59,6 +59,53 @@ def solve_ivp_for_terminal_value(*, f, df, tspan, u0, solver, **solver_kwargs):
     return extract_qoi_fn(t, state)
 
 
+@partial(jax.jit, static_argnames=("f", "df", "solver"))
+def solve_ivp_saveat(*, f, df, u0, solver, saveat, **solver_kwargs):
```
Comment (pnkraemer) [inline tornadox/solve.py:63]: if you have a better name than `solve_ivp_saveat`, please say so. I considered "checkpoint", but preferred `saveat` because it has less autodiff-connotation.

File: tornadox/solvers/ek1.py [hunk 4]
```diff
@@ -189,3 +189,232 @@ def extract_qoi_fn(t, state):
...
+        # Condense both backward models into one.
+        # (Technically speaking, this implements 'fixed-point smoothing')
+        # todo: if we do this somewhere else, perform_step_fn() can be reused
+        #  for other solve_ivp()-style functions.
+        A, (b, B_sqrtm) = init_val.backward_model
+        C, (d, D_sqrtm) = state.backward_model
+        g = A @ C
+        xi = A @ d + b
+        Xi = sqrtutil.sum_of_sqrtm_factors(S1=A @ D_sqrtm, S2=B_sqrtm)
+        backward_model_fp = (g, (xi, Xi))
```
Comment (pnkraemer) [inline tornadox/solvers/ek1.py:382]: If you don't get this (which would be very understandable! :D) ping me offline

File: tox.ini [hunk 5]
```diff
@@ -34,7 +34,7 @@ deps =
     jupyter
     jupytext
 commands =
-    jupytext --execute docs/tutorial_notebooks/*.ipynb
+    jupytext --execute docs/tutorial_notebooks/*.md
```
Comment (pnkraemer) [inline tox.ini:37]: that was accidentally reverted when I resolved a merge conflict earlier today. `.md` is the correct one.

File: tornadox/solvers/ek1.py (hunk 4 above)
Comment (nathanaelbosch) [inline tornadox/solvers/ek1.py:382]: I know we're not so consistent yet, but here I really find it helpful if we can have matrices capitals and vectors lower-case :sweat_smile: 
```suggestion
        G = A @ C
        xi = A @ d + b
        Xi = sqrtutil.sum_of_sqrtm_factors(S1=A @ D_sqrtm, S2=B_sqrtm)
        backward_model_fp = (G, (xi, Xi))
```

File: tornadox/solvers/ek1.py [hunk 6]
```diff
@@ -189,3 +189,232 @@ def extract_qoi_fn(t, state):
...
+            u=state.u,
+            backward_model=backward_model_fp,
+            dt_proposed=state.dt_proposed,
+            error_norm=state.error_norm,
+            stats=stats,
+        )
+        return t, state
+
+    @jax.jit
+    def reset_state_at_checkpoint_fn(t, state):
```
Comment (nathanaelbosch) [inline tornadox/solvers/ek1.py:398]: I do not have an intuition of what this is used for.

File: tornadox/solvers/ek1.py [hunk 7]
```diff
@@ -189,3 +189,232 @@ def extract_qoi_fn(t, state):
...
+
+    # Initialisation with autodiff
+    if num_derivatives <= 5:
+        autodiff_fun = autodiff_first_order.forwardmode_jvp
+    else:
+        autodiff_fun = autodiff_first_order.taylormode
+
+    # At each checkpoint, the backward model needs to be 'reset'.
+    # This means that it is set to an identity, noise-free transition.
+    # We just assemble this once and keep referencing it below.
```
Comment (nathanaelbosch) [inline tornadox/solvers/ek1.py:224]: Does this actually provide a benefit in jax? Since everything is static anyways I'd have assumed that it doesn't change much, and in that case just calling the identity_transition directly might make it more readable.

EDIT: Actually, maybe just renaming helps already. What do you think of making "identity" more prominent? Maybe it doesn't even need to be called "bw_model", since it is only used when assigned to keyword values `backward_model`?

File: tornadox/solvers/ek1.py (hunk 4 above)
Comment (nathanaelbosch) [reply tornadox/solvers/ek1.py:382]: Seems fine :) This does seem fairly modular though, it combines two linear models... how about making it a function already? :sweat_smile:

**Review by nathanaelbosch — CHANGES_REQUESTED** [review]
I only skimmed the tests, it's a bit hard to really understand the outputs at this point. I think I'll need a proper tutorial to fully grasp what we'll have^^ In any case, I made some comments that need clarification. Thanks already for the PR!

File: tornadox/solve.py (hunk 3 above)
Comment (pnkraemer) [reply tornadox/solve.py:63]: No, I dont think so. Same discussion as for the attempt_step() vs attempt_step_forward_only(): The saveat-versions are strictly more expensive per step. (And it is also much more complicated in terms of implementation, because it nests all sorts of loops)

File: tornadox/solvers/ek1.py (hunk 4 above)
Comment (pnkraemer) [reply tornadox/solvers/ek1.py:382]: where would this be reused?

File: tornadox/solve.py [hunk 8]
```diff
@@ -59,6 +59,53 @@ def solve_ivp_for_terminal_value(*, f, df, tspan, u0, solver, **solver_kwargs):
...
+    # assume t0 and tmax are the "boundary" elements in saveat.
+    # As such, `saveat` makes `tspan` obsolete
+    # Also assume it is sorted.
```
[the remark this answers is not in the document]
Comment (pnkraemer) [reply tornadox/solve.py:66]: Not sure, I think I don't like sorting the inputs internally. I could be convinced to have a check and an error message here at some point in the future, though :) 

What would you as a user expect to receive when inputting non-sorted saveat-points?. And isn't sorting usually somewhere between O(n) and O(n^2) in complexity (best case vs worst case), isn't it? And finally, when solving backwards, maybe the saveat points should be allowed to be decreasing.

File: tornadox/solvers/ek1.py (hunk 4 above)
Comment (pnkraemer) [reply tornadox/solvers/ek1.py:382]: I am not sure I am actually a fan of this capitalisation. But at this point here, I changed it in order to be consistent and to make it easier for you to read :)

File: tornadox/blocks/markov.py [hunk 9]
```diff
@@ -0,0 +1,21 @@
+"""Utilities for working with posteriors."""
+
+import jax
+
+from tornadox.blocks import sqrtutil
+
+
+def marginalize(*, m0, c_sqrtm0, A, b, Q_sqrtm, reverse):
+    """Compute the marginals of a linear model."""
```
[the remark this answers is not in the document]
Comment (pnkraemer) [reply tornadox/blocks/markov.py:9]: Marginalises a sequence of outputs from a linear state-space model. Therefore the scan. I updated the docstring.

File: tornadox/solve.py [hunk 10]
```diff
@@ -59,6 +59,53 @@ def solve_ivp_for_terminal_value(*, f, df, tspan, u0, solver, **solver_kwargs):
...
+    # If not, anything could happen.
+
+    # The signatures of those functions are the same as in solve_ivp_terminal_value
+    # with the exception of the additional reset_state_at_checkpoint_fn().
+    # It maps (t,state) to (t, state), and replaces some internals accordingly.
```
[the remark this answers is not in the document]
Comment (pnkraemer) [reply tornadox/solve.py:71]: I updated the comment.

File: tornadox/solve.py (hunk 3 above)
Comment (nathanaelbosch) [reply tornadox/solve.py:63]: Ah is it because a saveat implementation of the terminal value solver would then also attempt to smooth back all the way to the initial value? Didn't consider that, but I see that that would be more expensive.

File: tornadox/solvers/ek1.py (hunk 4 above)
Comment (nathanaelbosch) [reply tornadox/solvers/ek1.py]: I don't know. My motivation would have been readability, both of this block and then also of the surrounding stuff. I'd still say it would improve the code to make it a function, but feel free to decide as you prefer.

File: tornadox/solve.py (hunk 3 above)
Comment (pnkraemer) [reply tornadox/solve.py:63]: Not necessarily that (because the choice between smoothing, sampling, or doing something else is up to the user), but more so because it computes a bunch of smoothing gains that are not needed.

File: tornadox/solvers/ek1.py (hunk 7 above)
Comment (nathanaelbosch) [reply tornadox/solvers/ek1.py:224]: Cool, thanks!

File: tornadox/solve.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (pnkraemer) [reply tornadox/solve.py:66]: I added a "todo: tell user that sorting is important" comment to the code

File: tornadox/solvers/ek1.py (hunk 4 above)
Comment (pnkraemer) [reply tornadox/solvers/ek1.py]: I extracted this into a `condense_linear_model()` function.

File: tornadox/solvers/ek1.py [hunk 11]
```diff
@@ -374,14 +374,11 @@ def body_fun(s):
 
         # Condense both backward models into one.
         # (Technically speaking, this implements 'fixed-point smoothing')
-        # todo: if we do this somewhere else, perform_step_fn() can be reused
+        # todo: if we do this condensing somewhere else, perform_step_fn() can be reused
```
Comment (nathanaelbosch) [inline tornadox/solvers/ek1.py:377]: What does this comment mean actually?

File: tornadox/blocks/markov.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (nathanaelbosch) [reply tornadox/blocks/markov.py]: Still not 100% sure. I mean I think it's a sequence of predictions. A, b, and Q are not just one but many transitions, stacked along axis 0?

[the remark this answers is not in the document]
Comment (pnkraemer) [reply tornadox/blocks/markov.py]: But predictions are marginalisation, right?

[the remark this answers is not in the document]
Comment (pnkraemer) [reply tornadox/blocks/markov.py]: ah okay, I see. The function could probably use an updated docstring, I will try to get around to doing it in the next few days (which are pretty busy with other, unrelated stuff, unfortunately)

File: tornadox/solvers/ek1.py (hunk 11 above)
Comment (pnkraemer) [reply tornadox/solvers/ek1.py:377]: if the line: 
linear-model \circ linear-model = new-linear-model
is performed not inside perform_step_fn, but outside of that, then perform_step_fn can be reused as a default ode-solver-step, instead of being tied to the saveat/checkpoint version

Comment (nathanaelbosch) [reply tornadox/solvers/ek1.py:377]: I see, thanks
